lib/chef/provider.rb

```diff
@@ -175,6 +175,57 @@ def converge_by(descriptions, &block)
       converge_actions.add_action(descriptions, &block)
     end
 
+    #
+    # Handle patchy convergence safely.
+    #
+    # - Does *not* call the block if the current_resource's properties match
+    #   the properties the user specified on the resource.
+    # - Calls the block if current_resource does not exist
+    # - Calls the block if the user has specified any properties in the resource
+    #   whose values are *different* from current_resource.
+    # - Does *not* call the block if why-run is enabled (just prints out text).
+    # - Prints out automatic green text saying what properties have changed.
+    #
+    # @param properties An optional list of property names (symbols). If not
+    #   specified, `new_resource.class.state_properties` will be used.
+    # @param converge_block The block to do the converging in.
+    #
+    # @return [Boolean] whether the block was executed.
+    #
+    def converge_if_changed(*properties, &converge_block)
+      properties = new_resource.class.state_properties.map { |p| p.name } if properties.empty?
+      properties = properties.map { |p| p.to_sym }
+      if current_resource
+        # Collect the list of modified properties
+        specified_properties = properties.select { |property| new_resource.property_is_set?(property) }
+        modified = specified_properties.select { |p| new_resource.send(p) != current_resource.send(p) }
```

Comment:
I can imagine some scenarios where two different values mean the same thing (like `'777'` and `0777` for `mode`, and the user would have to fix with either coercion and/or a type with an equality operator (or both). Is that what you meant?

One thing to keep in mind is the implementor of the resource is in control of both `load_current_resource` and calling `converge_if_changed`. Another is that the above error manifests as updating when you don't need to, which sucks, but is a much better problem than its counterpart, *not* updating when you *should* :)

Still, I'm all ears for specifics--is that the scenario you had in mind as far as "improper load_current_resource"?